File: src/api/router.py
import logging
import time
from fastapi import APIRouter, HTTPException
from src.api.schemas import HealthResponse, SearchRequest
from src.ingestion.arxiv.client import ArxivClient
from src.ingestion.arxiv.models import ArxivPaper, ArxivSearchResult

# Configure logger
logger = logging.getLogger("research_copilot.api")

router = APIRouter(prefix="/api/v1", tags=["Literature Ingestion"])
arxiv_client = ArxivClient()


@router.get("/health", response_model=HealthResponse, summary="Health Check")
async def health_check():
    """Verify backend API operational status."""
    logger.info("Health check endpoint queried")
    return HealthResponse()


@router.post(
    "/search/arxiv",
    response_model=ArxivSearchResult,
    summary="Search arXiv Literature (POST)",
    description="Retrieve top-K relevant research papers using the HTTP POST method with JSON body.",
)
async def search_arxiv_query(req: SearchRequest):
    """POST method endpoint supporting standard HTTP POST method with body."""
    logger.info("Received arXiv search request via HTTP POST method. Query: '%s', Top K: %d", req.query, req.top_k)
    start_time = time.time()
    try:
        results = arxiv_client.search(
            query=req.query,
            max_results=req.top_k,
            sort_by=req.sort_by,
            sort_order=req.sort_order,
        )
        duration = time.time() - start_time
        logger.info(
            "Successfully fetched %d papers for query '%s' in %.2f seconds (total matched: %d)",
            results.returned_count,
            req.query,
            duration,
            results.total_results
        )
        return results
    except Exception as e:
        logger.error("Exception occurred while searching arXiv for query '%s': %s", req.query, str(e), exc_info=True)
        return ArxivSearchResult(
            query=req.query,
            total_results=0,
            returned_count=0,
            papers=[],
        )


@router.get(
    "/papers/arxiv/{arxiv_id:path}",
    response_model=ArxivPaper,
    summary="Fetch Paper Metadata by arXiv ID",
    description="Retrieve detailed metadata for a single arXiv paper by ID (e.g. 1706.03762).",
)
async def get_paper_by_id(arxiv_id: str):
    logger.info("Received request to fetch paper by ID: '%s'", arxiv_id)
    try:
        results = arxiv_client.search(query=f"id:{arxiv_id}", max_results=1)
        if not results.papers:
            logger.warning("Paper with arXiv ID '%s' not found.", arxiv_id)
            raise HTTPException(status_code=404, detail=f"Paper with arXiv ID '{arxiv_id}' not found.")
        logger.info("Successfully fetched paper details for ID: '%s'", arxiv_id)
        return results.papers[0]
    except HTTPException:
        raise
    except Exception as e:
        logger.error("Error occurred while retrieving paper '%s': %s", arxiv_id, str(e), exc_info=True)
        raise HTTPException(status_code=500, detail=f"Error retrieving paper '{arxiv_id}': {str(e)}")

File: src/api/router.go
package api

import (
	"encoding/json"
	"io"
	"log"
	"net/http"
	"os"
	"time"

	"research_copilot/src/core"
	"research_copilot/src/ingestion/arxiv"
	"research_copilot/src/ingestion/crossref"
	"research_copilot/src/ingestion/github"
	"research_copilot/src/ingestion/huggingface"
	"research_copilot/src/ingestion/kaggle"
	"research_copilot/src/ingestion/openalex"
	"research_copilot/src/ingestion/paperswithcode"
	"research_copilot/src/ingestion/pubmed"
	"research_copilot/src/ingestion/semanticscholar"
	"research_copilot/src/retrieval"
)

var arxivClient = arxiv.NewArxivClient()
var githubClient = github.NewGithubClient()
var hfClient = huggingface.NewHuggingFaceClient()
var s2Client = semanticscholar.NewS2Client()
var kaggleClient = kaggle.NewKaggleClient()
var openAlexClient = openalex.NewOpenAlexClient()
var crossrefClient = crossref.NewCrossrefClient()
var pwcClient = paperswithcode.NewPWCClient()
var pubmedClient = pubmed.NewPubMedClient()
var hybridEngine = retrieval.NewHybridEngine()

// RegisterRoutes registers the handlers on the given HTTP ServeMux
func RegisterRoutes(mux *http.ServeMux) {
	mux.HandleFunc("/api/v1/health", handleHealth)
	mux.HandleFunc("/api/v1/search/arxiv", handleSearchArxiv)
	mux.HandleFunc("/api/v1/papers/arxiv/", handleGetPaperByID)
	mux.HandleFunc("/api/v1/search/huggingface", handleSearchHuggingFace)
	mux.HandleFunc("/api/v1/search/semanticscholar", handleSearchSemanticScholar)
	mux.HandleFunc("/api/v1/search/kaggle", handleSearchKaggle)
	mux.HandleFunc("/api/v1/search/openalex", handleSearchOpenAlex)
	mux.HandleFunc("/api/v1/search/crossref", handleSearchCrossref)
	mux.HandleFunc("/api/v1/search/paperswithcode", handleSearchPapersWithCode)
	mux.HandleFunc("/api/v1/search/pubmed", handleSearchPubMed)
	mux.HandleFunc("/api/v1/search/unified", handleSearchUnified)
	mux.HandleFunc("/api/v1/retrieval/hybrid", handleHybridRetrieval)
	mux.HandleFunc("/api/v1/search/sessions", handleGetSearchSessions)
	mux.HandleFunc("/api/v1/papers/by-request/", handleGetPapersByRequestID)
	mux.HandleFunc("/api/v1/knowledge-graph", handleGetKnowledgeGraph)

	// Serve static files from public/ directory on root URL
	fs := http.FileServer(http.Dir("./public"))
	mux.Handle("/", fs)
}

func writeJSONResponse(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(data)
}

func writeJSONError(w http.ResponseWriter, status int, errMsg string) {
	writeJSONResponse(w, status, map[string]string{"error": errMsg})
}

func handleHealth(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeJSONError(w, http.StatusMethodNotAllowed, "Method Not Allowed")
		return
	}
	log.Println("[API] Health check endpoint queried")
	writeJSONResponse(w, http.StatusOK, HealthResponse{Status: "healthy"})
}

func handleGetKnowledgeGraph(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeJSONError(w, http.StatusMethodNotAllowed, "Method Not Allowed")
		return
	}

	kgPath := ".ua/knowledge-graph.json"
	data, err := os.ReadFile(kgPath)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Knowledge graph file not found")
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(data)
}

func handleSearchPubMed(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeJSONError(w, http.StatusMethodNotAllowed, "Method Not Allowed")
		return
	}

	var req SearchRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSONError(w, http.StatusBadRequest, "Invalid JSON body: "+err.Error())
		return
	}

	if req.Query == "" {
		writeJSONError(w, http.StatusBadRequest, "Missing query parameter")
		return
	}

	if req.TopK <= 0 {
		req.TopK = 5
	}

	res, err := pubmedClient.Search(r.Context(), req.Query, req.TopK)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "PubMed API search failed: "+err.Error())
		return
	}

	if err := pubmed.IngestPubMedSearchResult(r.Context(), core.DB, res); err != nil {
		log.Printf("[API] Failed to ingest PubMed search results into DB: %v", err)
	}

	writeJSONResponse(w, http.StatusOK, res)
}

func handleHybridRetrieval(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeJSONError(w, http.StatusMethodNotAllowed, "Method Not Allowed")
		return
	}

	// Proxy request to Python Agentic RAG Retrieval Service (Port 8104)
	client := &http.Client{Timeout: 30 * time.Second}
	req, err := http.NewRequestWithContext(r.Context(), "POST", "http://localhost:8104/retrieval/hybrid", r.Body)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to proxy retrieval request: "+err.Error())
		return
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		writeJSONError(w, http.StatusServiceUnavailable, "Python Retrieval Service unreachable: "+err.Error())
		return
	}
	defer resp.Body.Close()

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(resp.StatusCode)
	_, _ = io.Copy(w, resp.Body)
}



